tests/TestAssets/test-suite-data.json

```diff
@@ -100,8 +100,8 @@
   },
   {
     "description": "docker uses qualifiers and hash image id as versions",
-    "purl": "pkg:docker/customer/dockerimage@sha256:244fd47e07d1004f0aed9c?repository_url=gcr.io",
-    "canonical_purl": "pkg:docker/customer/dockerimage@sha256:244fd47e07d1004f0aed9c?repository_url=gcr.io",
+    "purl": "pkg:docker/customer/dockerimage@sha256%3A244fd47e07d1004f0aed9c?repository_url=gcr.io",
+    "canonical_purl": "pkg:docker/customer/dockerimage@sha256%3A244fd47e07d1004f0aed9c?repository_url=gcr.io",
```

Comment:
@pombredanne according to the [spec text](https://github.com/package-url/purl-spec/blob/c02b002f09bdc88a501f62259eec18761957828a/PURL-SPECIFICATION.rst#character-encoding):

> * the '#', '?', '@' and ':' characters must NOT be encoded when used as separators. <ins>They may need to be encoded elsewhere</ins>
> * the ':' scheme and type separator does not need to and must NOT be encoded. It is unambiguous unencoded everywhere

here ':' is part of `version` and not separator. Should we interpret the `They may need to be encoded elsewhere` as `it is optional to encode them` since both [Python](https://github.com/package-url/packageurl-python/blob/b0111ad7e559b996f6ec725cd803e95ad142d2a8/tests/data/test-suite-data.json#L89) and [JavaScript](https://github.com/package-url/packageurl-js/blob/3e0e10616b23ba5c1a9529f31357abd5ee0bfa72/test/data/test-suite-data.json#L113) implementations honor the un-encoded form?

Reply:
Huh - interesting point. It looks like both the Python and Javascript implementations simply call `.Replace("%3A", ":")` after percent-encoding the `name`, `version`, and `subpath` components.

Looking at qualifier values, [the Javascript implementation also allows to be unencoded `/`, but still encodes `:`](https://github.com/package-url/packageurl-js/blob/3e0e10616b23ba5c1a9529f31357abd5ee0bfa72/src/package-url.js#L110), whereas the Python implemetation [does the opposite for qualifier values](https://github.com/package-url/packageurl-python/blob/b0111ad7e559b996f6ec725cd803e95ad142d2a8/src/packageurl/__init__.py#L227) - so opinions seem to vary there.

@morended What do you think about using `.Replace("%3A", ":")` to allow `:` to stay unencoded in the `name`, `version` and `subpath` components (to match the other implementations)? We could punt on `/` and `:` in qualifier values for now since there isn't a consensus in those implementations, and leaving special characters unencoded is optional after all.

Reply:
@am11 I have addressed this in my latest revision. Let me know if there is anything.